crates/typst-library/src/layout/flow.rs

```diff
@@ -436,6 +440,24 @@ impl<'a> FlowLayouter<'a> {
 
     /// Finish the frame for one region.
     fn finish_region(&mut self, vt: &mut Vt) -> SourceResult<()> {
+        if let Some(items) = self.last_frame_only_out_of_flow.take() {
+            self.items.splice(0..0, items);
+            self.finished.pop();
+            self.regions.next();
+            self.initial = self.regions.size;
+            self.has_footnotes = false;
+        }
+
+        if !self.items.is_empty() && self.items.iter().all(|item| {
+            match item {
+                FlowItem::Placed {float: false, ..} => true,
+                FlowItem::Frame {frame, ..} => frame.size().is_zero(),
+                _ => false,
+            }
+        }) {
+            self.last_frame_only_out_of_flow = Some(self.items.clone());
```

Comment:
I had to make `FlowItem` Clone, which is a bit sad. It should be noted, though, that this clone only happens when the last frame only contains out-of-flow elements, which is not often.

Reply:
> Or wait, is that why you're pop()-ing a frame above?

Yes, in case the last frame was out-of-flow, we pop it from the list of finished frames before adding its content (which we cloned here) to the next frame.